pgctl/service.py

```diff
@@ -215,17 +215,53 @@ def background(self):
             return
 
         with self.flock() as lock:
-            log = self.path.join('log').open('a')
-            log = prepend_timestamps_to(log)
+            log_fifo_path = self.path.join('log_pipe').strpath
+            if not os.path.exists(log_fifo_path):
+                os.mkfifo(log_fifo_path)
+
+            # The writer (opened as RW to avoid blocking) must be opened
+            # before checking if the logging process is running,
+            # since we're guaranteed* that if a writing pipe is open,
+            # and the logger is running, it will remain running.
+            #
+            # Thus, we can maintain the assertion that if we find the logger
+            # is running, it will remain running.
+            #
+            # * assuming the process isn't manually killed
+            log_fifo_writer = os.open(log_fifo_path, os.O_RDWR)
+            if not self.is_logger_running():
+
+                with open(os.devnull, 'w') as devnull:
+                    log_fifo_reader = os.open(log_fifo_path, os.O_RDONLY)
```

Comment:
The forking version is racier because we don't have a pipe open before we check on the state of the logging process.